ironfish-rust/src/public_account/description.rs

```diff
@@ -0,0 +1,275 @@
+        writer.write_all(&signers_len.to_le_bytes())?;
+
+        for signer in &self.signers {
+            writer.write_all(signer.as_bytes())?;
+        }
+
+        let signatures_len: u16 = self.signatures.len().try_into()?;
+        writer.write_all(&signatures_len.to_le_bytes())?;
+
+        for signature in &self.signatures {
+            writer.write_all(&signature.to_bytes())?;
+        }
+
+        let transfers_len: u16 = self.transfers.len().try_into()?;
+        writer.write_all(&transfers_len.to_le_bytes())?;
+
+        for transfer in &self.transfers {
+            transfer.write(&mut writer)?;
+        }
+
+        self.address.write(&mut writer)?;
+
+        Ok(())
+    }
+
+    fn valid(&self) -> Result<(), IronfishError> {
+        if self.min_signers < 1 {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signers.len() < self.min_signers as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        let unique_signers: Vec<_> = self.signers.iter().collect();
+        if unique_signers.len() != self.signers.len() {
+            return Err(IronfishError::new(IronfishErrorKind::DuplicateSigner));
+        }
+
+        let unique_signatures: Vec<_> = self.signatures.iter().collect();
+        if unique_signatures.len() != self.signatures.len() {
+            return Err(IronfishError::new(IronfishErrorKind::DuplicateSignature));
+        }
+
+        Ok(())
+    }
+
+    fn verify(&self) -> Result<(), IronfishError> {
+        self.valid()?;
+
+        let hash = &self.hash()?;
+        for signature in &self.signatures {
+            let is_valid = self
+                .signers
+                .iter()
+                .any(|signer| signer.verify(hash, signature).is_ok());
+
+            if !is_valid {
+                return Err(IronfishError::new(IronfishErrorKind::InvalidSignature));
+            }
+        }
+        Ok(())
+    }
+
+    pub fn hash(&self) -> Result<[u8; 32], IronfishError> {
+        // TODO(jwp): verify which hashers supported by axelar
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&self.version.to_le_bytes());
+        hasher.update(&self.min_signers.to_le_bytes());
+        let signer_len = self.signers.len() as u16;
+        hasher.update(&signer_len.to_le_bytes());
+        for signer in &self.signers {
+            hasher.update(signer.as_bytes());
+        }
+        let transfer_len = self.transfers.len() as u16;
+        hasher.update(&transfer_len.to_le_bytes());
+        for transfer in &self.transfers {
+            hasher.update(&transfer.as_bytes()?);
+        }
+        hasher.update(&self.address.public_address());
+        Ok(hasher.finalize().into())
+    }
+
+    pub fn add_signatures(&mut self, signatures: Vec<Signature>) -> Result<(), IronfishError> {
+        self.signatures.extend(signatures);
+        self.verify()
```

Comment:
Note that if the verification fails, `signatures` will still include the invalid signature. We should make sure that `signatures` does not get modified if it's invalid.

I suggest changing `verify(&self)` to `verify_signature(&self, signature: &Signature)` (or similar) and calling it before updating `self.signatures`